File: commons-levelops/database-commons/src/main/java/io/levelops/commons/databases/models/database/Integration.java
package io.levelops.commons.databases.models.database;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonValue;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.ToString;
import org.apache.commons.collections4.MapUtils;
import org.apache.commons.lang3.EnumUtils;

import javax.annotation.Nullable;
import java.util.List;
import java.util.Map;

@Getter
@NoArgsConstructor
@AllArgsConstructor
@EqualsAndHashCode
@ToString
@Builder(toBuilder = true)
@JsonInclude(JsonInclude.Include.NON_NULL)
public class Integration {
    @JsonProperty("tags")
    private List<String> tags;

    @JsonProperty("id")
    private String id;

    @JsonProperty("name")
    private String name;

    @JsonProperty("description")
    private String description;

    @JsonProperty("url")
    private String url;

    @JsonProperty("status")
    private String status;

    @JsonProperty("application")
    private String application;

    @JsonProperty("satellite")
    private Boolean satellite;

    @JsonProperty("updated_at")
    private Long updatedAt;

    @JsonProperty("created_at")
    private Long createdAt;

    @JsonProperty("metadata")
    private Map<String, Object> metadata;

    @JsonProperty("append_metadata")
    private boolean appendMetadata;

    /**
     * If the integration supports multiple authentication schemes, this represents the one that was selected.
     */
    @JsonProperty("authentication")
    Authentication authentication;

    @JsonProperty("linked_credentials")
    private String linkedCredentials; // (optional) integration id; if specified, this integration will use the credentials of that integration instead

    public enum Authentication {
        UNKNOWN, // leaving this for migration purposes
        NONE, // no authentication needed or provided
        OAUTH,
        API_KEY,
        MULTIPLE_API_KEYS,
        DB,
        ATLASSIAN_CONNECT_JWT,
        ADFS;

        @JsonCreator
        @Nullable
        public static Authentication fromString(@Nullable String value) {
            return EnumUtils.getEnumIgnoreCase(Authentication.class, value);
        }

        @JsonValue
        @Override
        public String toString() {
            return super.toString().toLowerCase();
        }
    }

    @JsonIgnore
    public boolean isGithubAppsIntegration() {
        if (application.equals("github")) {
            Map<String, Object> metadata = MapUtils.emptyIfNull(getMetadata());
            return metadata.containsKey("app_id");
        }
        return false;
    }
}

File: ui-levelops/src/model/entities/Integration.ts
import { Entity } from "./entity";

export type AuthenticationTypes = "oauth" | "none" | "api_key" | "multiple_api_keys";
export type IntegrationMetadataKeys =
  | "is_push_based"
  | "fetch_commits"
  | "fetch_prs"
  | "fetch_issues"
  | "fetch_projects"
  | "fetch_commit_files"
  | "fetch_action_logs"
  | "repos"
  | "subtype";

export interface Integration extends Entity {
  append_metadata: boolean;
  application: string;
  authentication: AuthenticationTypes;
  created_at: number;
  description: string;
  metadata: Record<IntegrationMetadataKeys, boolean | string>;
  name: string;
  satellite: boolean;
  status: string;
  tags: Array<string>;
  updated_at: number;
  url: string;
}

export enum AzureIntegrationSubType {
  WI = "wi",
  SCM = "scm",
  CICD = "cicd"
}

export enum AzureApplicationSubType {
  BOARDS = "azure_devops_boards",
  PIPELINES = "azure_devops_pipelines",
  REPOS = "azure_devops_repos"
}
